src/main/ml-modules/root/com.marklogic.smart-mastering/impl/process.xqy

```diff
@@ -95,6 +101,34 @@ declare function proc-impl:process-match-and-merge-with-options(
           matcher:save-match-notification($label, ($uri, $notify-uris))
         },
         map:new((map:entry("isolation", "different-transaction"), map:entry("update", "true")))
+      ),
+
+    let $actions := $matching-results/result/@action[fn:not(. = $const:NOTIFY-ACTION or . = $const:MERGE-ACTION)]
+    for $action in fn:distinct-values($actions)
+    let $action-xml := $matching-options//*:action[@name = $action]
+    let $action-func :=
+      fun-ext:function-lookup(
+        fn:string($action-xml/@function),
+        fn:string($action-xml/@namespace),
+        fn:string($action-xml/@at),
+        ()
       )
+    let $filtered-matches := $matching-results/result[@action = $action]
```

Comment:
@dmcassel @paxtonhare IIUC - this allows for a developer to provide a custom implementation for merging, correct? e.g. merge entity B into entity A, preserving the ID of entity A and not creating a new entity?

Reply:
yes, it allows for an arbitrary action. if someone implemented what you're suggesting, I wouldn't be confident about unmerging or auditing working as expected. The intent of this is to provide additional notification actions (send me an email) as opposed to changing the basic way things work, but that is a possibility